docs/storage-drivers/s3.md

```diff
@@ -29,6 +29,55 @@ An implementation of the `storagedriver.StorageDriver` interface which uses Amaz
 
 `v4auth`: (optional) Whether you would like to use aws signature version 4 with your requests. This defaults to false if not specified (note that the eu-central-1 region does not work with version 2 signatures, so the driver will error out if initialized with this region and v4auth set to false)
 
-`chunksize`: (optional) The default part size for multipart uploads (performed by WriteStream) to s3. The default is 10 MB. Keep in mind that the minimum part size for s3 is 5MB. You might experience better performance for larger chunk sizes depending on the speed of your connection to s3.
+`chunksize`: (optional) The default part size for multipart uploads (performed by WriteStream) to S3. The default is 10 MB. Keep in mind that the minimum part size for S3 is 5MB. You might experience better performance for larger chunk sizes depending on the speed of your connection to S3.
 
 `rootdirectory`: (optional) The root directory tree in which all registry files will be stored. Defaults to the empty string (bucket root).
+
+# CloudFront as Middleware /w S3 backend
```

Comment:
@djenriquez Nice contribution. I had a few tweaks to the language and head levels.  

```markdown
## CloudFront as Middleware /w S3 backend

Adding CloudFront as a middleware for your S3 backed registry can dramatically improve pull times. By using Cloudfront, you registry will have the ability to retrieve your images from edge servers, rather than the geographically-limited location of your S3 bucket. The farther your registry is from your bucket, the more improvements you will see. See [Amazon CloudFront](https://aws.amazon.com/cloudfront/details/).

## Configuring CloudFront for Distribution

If you are unfamiliar with creating a CloudFront distribution, see [Getting
Started with
Cloudfront](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/GettingStarted.html).

Defaults can be kept in most areas except origin, behaviors, and in the registory configuration.

### Origin

Create your CloudFront distribution such that the `Origin Path` is set to the directory level of the root "docker" key in S3. If your registry exists on the bucket's root, leave this path blank.

### Behaviors

    Set the following:
    
    - Viewer Protocol Policy: HTTPS Only
    - Allowed HTTP Methods: GET, HEAD, OPTIONS, PUT, POST, PATCH, DELETE
    - Cached HTTP Methods: OPTIONS (checked)
    - Restrict Viewer Access (Use Signed URLs or Signed Cookies): Yes
      - Trusted Signers: Self (Can add other accounts as long as you have access to CloudFront Key Pairs for those additional accounts)

### Registry configuration

Make sure you use the `middleware` option. Make sure you keep the `storage` option because CloudFront can only handle `pull` actions. Any `push` actions are still directly written to S3. The following example shows what you will need at minimum:

```
...
storage:
  s3:
    region: us-east-1
    bucket: docker.myregistry.com
middleware:
  storage:
    - name: cloudfront
      options:
        baseurl: https://abcdefghijklmn.cloudfront.net/
        privatekey: /etc/docker/cloudfront/pk-ABCEDFGHIJKLMNOPQRST.pem
        keypairid: ABCEDFGHIJKLMNOPQRST
...
```

## CloudFront Key-Pair

A CloudFront key-pair is required for all AWS accounts needing access to your
CloudFront distribution. For information, please see [Creating CloudFront Key
Pairs](http://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/private-content-trusted-signers.html#private-content-creating-cloudfront-key-pairs).
```